Error when using fully async + masking off policy + partial rollout

There is an error when running Qwen30A3 with fully async + partial rollout:
```
[36m(RolloutManager pid=2616346)[0m [2026-01-17 12:12:48] rollout.py:251 - Final collected 16 samples from rollout to train
[36m(RolloutManager pid=2616346)[0m [2026-01-17 12:12:48] rollout.py:720 - perf 6: {'rollout/response_len/mean': 6342.875, 'rollout/response_len/median': 3439.0, 'rollout/response_len/max': 16384, 'rollout/response_len/min': 0, 'rollout/zero_std/count_0': 1, 'rollout/repetition_frac': 0.0, 'rollout/truncated_ratio': 0.625, 'perf/rollout_time': 108.95538687705994, 'perf/tokens_per_gpu_per_sec': 131.05478681941523, 'perf/longest_sample_tokens_per_sec': 150.37347367218223, 'perf/effective_tokens_per_gpu_per_sec': 58.21534099233659, 'perf/longest_effective_sample_tokens_per_sec': 150.37347367218223}
Traceback (most recent call last):
  File "/mnt/llm-train/users/explore-train/qingyu/slime/train_async.py", line 76, in <module>
    train(args)
  File "/mnt/llm-train/users/explore-train/qingyu/slime/train_async.py", line 64, in train
    rollout_data_curr_ref = ray.get(x) if (x := rollout_data_next_future) is not None else None
                            ^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/ray/_private/auto_init_hook.py", line 22, in auto_init_wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/ray/_private/client_mode_hook.py", line 104, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/ray/_private/worker.py", line 2967, in get
    values, debugger_breakpoint = worker.get_objects(
                                  ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/ray/_private/worker.py", line 1015, in get_objects
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AssertionError): [36mray::RolloutManager.generate()[39m (pid=2616346, ip=11.251.10.151, actor_id=f11859e05a72c27bbca001c402000000, repr=<slime.ray.rollout.RolloutManager object at 0x7f9911c1e5a0>)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/llm-train/users/explore-train/qingyu/slime/slime/ray/rollout.py", line 138, in generate
    data = self._convert_samples_to_train_data(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/llm-train/users/explore-train/qingyu/slime/slime/ray/rollout.py", line 362, in _convert_samples_to_train_data
    len(sample.loss_mask) == sample.response_length
AssertionError: loss mask length 8563 != response length 16384
Traceback (most recent call last):
  File "/mnt/llm-train/users/explore-train/qingyu/slime/train_async.py", line 76, in <module>
    train(args)
  File "/mnt/llm-train/users/explore-train/qingyu/slime/train_async.py", line 64, in train
    rollout_data_curr_ref = ray.get(x) if (x := rollout_data_next_future) is not None else None
                            ^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/ray/_private/auto_init_hook.py", line 22, in auto_init_wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/ray/_private/client_mode_hook.py", line 104, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/ray/_private/worker.py", line 2967, in get
    values, debugger_breakpoint = worker.get_objects(
                                  ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/ray/_private/worker.py", line 1015, in get_objects
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AssertionError): [36mray::RolloutManager.generate()[39m (pid=2616346, ip=11.251.10.151, actor_id=f11859e05a72c27bbca001c402000000, repr=<slime.ray.rollout.RolloutManager object at 0x7f9911c1e5a0>)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/llm-train/users/explore-train/qingyu/slime/slime/ray/rollout.py", line 138, in generate
    data = self._convert_samples_to_train_data(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/llm-train/users/explore-train/qingyu/slime/slime/ray/rollout.py", line 362, in _convert_samples_to_train_data
    len(sample.loss_mask) == sample.response_length
AssertionError: loss mask length 8563 != response length 16384
[1;34mwandb[0m: 
[1;34mwandb[0m: You can sync this run to the cloud by running:
[1;34mwandb[0m: [1mwandb sync /root/wandb/offline-run-20260117_120032-u6dnk55q[0m
[1;34mwandb[0m: Find logs at: [1;35mwandb/offline-run-20260117_120032-u6dnk55q/logs[0m
[36m(SGLangEngine pid=72732, ip=11.251.90.10)[0m [2026-01-17 12:12:48 TP0] Prefill batch, #new-seq: 1, #new-token: 114, #cached-token: 56, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[36m(SGLangEngine pid=72732, ip=11.251.90.10)[0m [2026-01-17 12:12:44] INFO:     11.251.10.151:60856 - "GET /health HTTP/1.1" 200 OK[32m [repeated 7x across cluster][0m
[36m(SGLangEngine pid=72733, ip=11.251.90.10)[0m [2026-01-17 12:12:48 TP0] Decode batch, #running-req: 2, #token: 11197, token usage: 0.01, cuda graph: True, gen throughput (token/s): 224.74, #queue-req: 0, [32m [repeated 91x across cluster][0m[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.

[36m(SGLangEngine pid=72732, ip=11.251.90.10)[0m [2026-01-17 12:12:48 TP0] Prefill batch, #new-seq: 1, #new-token: 114, #cached-token: 56, token usage: 0.00, #running-req: 1, #queue-req: 0, [32m [repeated 7x across cluster][0m
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.

[31m---------------------------------------[39m
[31mJob 'raysubmit_vT6udeggimcc3vUg' failed[39m
[31m---------------------------------------[39m

Status message: Job entrypoint command failed with exit code 1, last available logs (truncated to 20,000 chars):
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.
[2026-01-17 12:12:49] selector_events.py:1068 - socket.send() raised exception.

```
my running scripts:
```
 #!/bin/bash

# ---------------------------- kill them ---------------------------

pkill -9 -f "sglang::scheduler"
pkill -9 -f "ray::MegatronTrainRayActor"

# ---------------------------- nvlink check ---------------------------

NVLINK_COUNT=$(nvidia-smi topo -m 2>/dev/null | grep -o 'NV[0-9][0-9]*' | wc -l)
if [ "$NVLINK_COUNT" -gt 0 ]; then
    HAS_NVLINK=1
else
    HAS_NVLINK=0
fi
echo "HAS_NVLINK: $HAS_NVLINK (detected $NVLINK_COUNT NVLink references)"

# ---------------------------- system config ---------------------------

export PYTHONBUFFERED=16 # will prevent ray from buffering stdout/stderr
export WANDB_MODE=offline # offline wandb

PROJECT_DIR=/mnt/llm-train/users/explore-train/qingyu/slime # /root/slime is mounted from the docker env
SCRIPT_DIR=${PROJECT_DIR}/scripts # where is the scripts
LOCAL_IP=$(hostname -I | awk '{print $1}') # get master node local ip for submitting
TIMESTAMP=$(date +%Y%m%d_%H%M%S) # timestamp for naming
SAVE_DIR=/mnt/llm-train/users/explore-train/qingyu/PeRL/outputs/${TIMESTAMP}_gspo_qwen30ba3b
LOG_DIR=$SAVE_DIR/output.log # where to save log

source "${SCRIPT_DIR}/models/qwen3-30B-A3B.sh" # load model args from slime for megatron
mkdir -p ${SAVE_DIR} # create save dir

# ---------------------------- reward server ---------------------

REWARD_SERVER_IP=http://11.251.78.226:8123/reward

# ---------------------------- ray env ---------------------------

# Build the runtime environment JSON with proper variable substitution
# we need to offline wandb
RUNTIME_ENV_JSON="{
  \"env_vars\": {
    \"PYTHONPATH\": \"/root/Megatron-LM/\",
    \"CUDA_DEVICE_MAX_CONNECTIONS\": \"1\",
    \"NCCL_NVLS_ENABLE\": \"${HAS_NVLINK}\",
    \"RAY_ENABLE_OPENTELEMETRY\": \"0\",
    \"RAY_DISABLE_METRICS_COLLECTION\": \"1\",
    \"WANDB_MODE\": \"offline\"
  }
}"

# ---------------------------- running config ---------------------------

CKPT_ARGS=(
   --hf-checkpoint /mnt/llm-train/users/explore-train/zhangyuqi60/Nomerge/ms-swift/hf_outputs/qwen3-30b-s1-0103/
   --ref-load /mnt/llm-train/users/explore-train/qingyu/.cache/qwen3-30b-s1-0103_torch_dist/
   --load /mnt/llm-train/users/explore-train/qingyu/.cache/qwen3-30b-s1-0103_torch_dist/
   --save ${SAVE_DIR}
   --save-interval 32
)

ROLLOUT_ARGS=(
   # training data path
   --prompt-data /mnt/llm-train/users/explore-train/qingyu/.cache/dapo-math-17k/dapo-math-17k-rm.jsonl

   --input-key prompt
   --label-key label
   --apply-chat-template
   --rollout-shuffle

   # reward settings (使用 remote reward server)
   --rm-type remote_rm
   --rm-url $REWARD_SERVER_IP
   --num-rollout 4096

   # rollout settings
   --over-sampling-batch-size 64
   --rollout-batch-size 2
   --n-samples-per-prompt 8
   --rollout-max-response-len 16384
   --rollout-temperature 1
   --global-batch-size 16
   --balance-data

   # async rl settings
   --partial-rollout # aka pipelineRL
   --rollout-function-path examples.fully_async.fully_async_rollout.generate_rollout_fully_async
   --dynamic-sampling-filter-path slime.rollout.filter_hub.dynamic_sampling_filters.check_reward_nonzero_std
   --mask-offpolicy-in-partial-rollout # only use on policy data for training

   # routing replay
   --use-rollout-routing-replay
   --use-slime-router # need to bypass sgl-router https://github.com/THUDM/slime/issues/1145
                     # because it does not support returning routed experts

   --rollout-num-gpus 16 # total rollout gpus i.e., nnodes * per_node_gpus
   --actor-num-gpus-per-node 8
   --rollout-num-gpus-per-engine 2 # tp size for sglang
   # --sglang-dp-size 8
   # --sglang-enable-dp-attention
)

EVAL_ARGS=(
   --eval-interval 32
   --eval-prompt-data aime /mnt/llm-train/users/explore-train/qingyu/.cache/aime-2024/aime-2024.jsonl
   --n-samples-per-eval-prompt 16
   --eval-max-response-len 32768
   --eval-top-p 1
)

PERF_ARGS=(
   # actor gpu allocations
   --actor-num-nodes 2

   --tensor-model-parallel-size 2
   --sequence-parallel
   --pipeline-model-parallel-size 1
   --context-parallel-size 1
   --expert-model-parallel-size 8
   --expert-tensor-parallel-size 1

   --recompute-granularity full
   --recompute-method uniform
   --recompute-num-layers 1

   # --micro-batch-size 1
   --use-dynamic-batch-size
   --max-tokens-per-gpu 20480

   # default dropout in megatron is 0.1
   --attention-dropout 0.0
   --hidden-dropout 0.0

   # should be good for model performance
   --accumulate-allreduce-grads-in-fp32
   --attention-softmax-in-fp32

   # need to comment this when using model with MLA
   --attention-backend flash
)

GRPO_ARGS=(
   --advantage-estimator gspo
   # --use-kl-loss
   --kl-loss-coef 0.00 # no kl loss
   --kl-loss-type low_var_kl
   --entropy-coef 0.00 # no entropy loss
   --eps-clip 3e-4
   --eps-clip-high 4e-4 # clip ratio higher
   # ref: verl gspo
)

OPTIMIZER_ARGS=(
   --optimizer adam
   --lr 1e-6
   --lr-decay-style constant
   --weight-decay 0.1
   --adam-beta1 0.9
   --adam-beta2 0.98

   # offloading to cpu
#    --optimizer-cpu-offload
#    --overlap-cpu-optimizer-d2h-h2d
#    --use-precision-aware-optimizer
)

WANDB_ARGS=(
   --use-wandb
   --wandb-project slime
   --wandb-group gspo-qwen30ba3b
   --wandb-key "EMPTYEMPTYEMPTYEMPTYEMPTYEMPTYEMPTYEMPTYEMPTY" # offline saving logs
)

SGLANG_ARGS=(
   --sglang-mem-fraction-static 0.8
   --sglang-cuda-graph-bs 1 2 4 8 $(seq 16 8 256)
#    --sglang-enable-ep-moe
)

# ---------------------------- submit task ---------------------------

ray job submit --address="http://${LOCAL_IP}:8265" \
   --runtime-env-json="${RUNTIME_ENV_JSON}" \
   -- python3 ${PROJECT_DIR}/train_async.py \
   ${MODEL_ARGS[@]} \
   ${CKPT_ARGS[@]} \
   ${ROLLOUT_ARGS[@]} \
   ${OPTIMIZER_ARGS[@]} \
   ${GRPO_ARGS[@]} \
   ${WANDB_ARGS[@]} \
   ${PERF_ARGS[@]} \
   ${EVAL_ARGS[@]} \
   ${SGLANG_ARGS[@]} 2>&1 | tee ${LOG_DIR}
```

After checking for the source code, I found that this problem may be because of the loss mask settings:

sglang_rollout.py:197-199
```
197    # mask previous off-policy generation for partial rollout
198    if args.partial_rollout and args.mask_offpolicy_in_partial_rollout and sample.response_length > 0:
199        sample.loss_mask = [0] * sample.response_length
```

The partial rollout will mask the off-policy part, but there will be new tokens and `response_length` will be longer. But the loss mask is not updated.

```
168        sample.tokens = sample.tokens + new_response_tokens
169        sample.response_length += len(new_response_tokens)
170        sample.response += output["text"]
```

I think this is the direct cause of the error raised from this assert:

```
359            assert (
360                len(sample.loss_mask) == sample.response_length
361            ), f"loss mask length {len(sample.loss_mask)} != response length {sample.response_length}"

```

--- a/slime/rollout/sglang_rollout.py
+++ b/slime/rollout/sglang_rollout.py
@@ -169,6 +169,11 @@ async def generate(args: Namespace, sample: Sample, sampling_params: dict[str, A
         sample.response_length += len(new_response_tokens)
         sample.response += output["text"]
 
+        # When partial rollout and masking off policy is enabled, update the loss mask
+        if sample.loss_mask is not None:
+            assert args.partial_rollout and args.mask_offpolicy_in_partial_rollout
+            sample.loss_mask += [1] * len(new_response_tokens)
+
         if sample.rollout_log_probs is None:
             sample.rollout_log_probs = []
         sample.rollout_log_probs += new_response_log_probs

